File: crates/monty-js/ts/worker/dumpSign.ts
// HMAC signing of dump payloads for the wasm worker path, mirroring the
// native pool's host-side signing so dumps stay byte-portable between the two
// backends when the same key is supplied.
//
// `crates/monty-pool/src/dump_sign.rs` is the source of truth for the format
// and the error strings — this file must stay byte-compatible with it:
//
//   signed_dump := [version u8 = 0x01][tag: 32-byte HMAC-SHA256(key, CONTEXT || inner)][inner]
//
// Uses WebCrypto (async, constant-time verify), which both Node and
// secure-context browsers provide as `crypto.subtle`.

/** Minimum accepted dump-key length, matching `MIN_DUMP_KEY_LEN` in Rust. */
export const MIN_DUMP_KEY_LEN = 16

/** Format version prepended to every signed dump. */
const SIGNED_DUMP_VERSION = 0x01

/** Domain-separation prefix mixed into every MAC. */
const CONTEXT = new TextEncoder().encode('monty-dump-sign-v1')

/** HMAC-SHA256 output length in bytes. */
const TAG_LEN = 32

/** Generates a random 32-byte ephemeral dump key (pool-local dumps). */
export function generateDumpKey(): Uint8Array {
  // requireSubtle so a crypto-less environment gets the clear WebCrypto error
  // here rather than a bare ReferenceError on `crypto` (and a key generated in
  // a non-secure context would be unusable anyway — import needs `subtle`)
  requireSubtle()
  return crypto.getRandomValues(new Uint8Array(32))
}

/** Coerces a dump key — raw bytes, or a string (UTF-8-encoded, matching the
 *  Python binding's `dump_key: str`) — to bytes. */
export function dumpKeyBytes(key: string | Uint8Array): Uint8Array {
  return typeof key === 'string' ? new TextEncoder().encode(key) : key
}

/**
 * Imports a dump key — raw bytes, or a string (UTF-8-encoded) — as a WebCrypto
 * HMAC-SHA-256 key, rejecting keys shorter than [`MIN_DUMP_KEY_LEN`] bytes.
 * Throws when `crypto.subtle` is unavailable (a browser page not in a secure
 * context) — dump/load needs it; execution does not.
 */
export function importDumpKey(key: string | Uint8Array): Promise<CryptoKey> {
  const bytes = dumpKeyBytes(key)
  if (bytes.length < MIN_DUMP_KEY_LEN) {
    throw new Error(`dump key must be at least ${MIN_DUMP_KEY_LEN} bytes`)
  }
  return requireSubtle().importKey('raw', copyBytes(bytes), { name: 'HMAC', hash: 'SHA-256' }, false, [
    'sign',
    'verify',
  ])
}

/**
 * A memoizing lazy dump-key provider: validates `key` eagerly (so a short
 * key fails at pool creation) but defers key generation and the WebCrypto
 * import to the first call — pools that never dump touch no crypto, and a
 * non-secure browser context only fails on actual dump/load use.
 */
export function lazyDumpKey(key?: string | Uint8Array): () => Promise<CryptoKey> {
  // snapshot the key bytes now — the import is deferred, and a caller mutating
  // its Uint8Array after pool creation must not change the pool's signing key
  const bytes = key === undefined ? undefined : copyBytes(dumpKeyBytes(key))
  if (bytes !== undefined && bytes.length < MIN_DUMP_KEY_LEN) {
    throw new Error(`dump key must be at least ${MIN_DUMP_KEY_LEN} bytes`)
  }
  let imported: Promise<CryptoKey> | undefined
  return () => (imported ??= importDumpKey(bytes ?? generateDumpKey()))
}

/** Signs a worker dump envelope, prepending the version byte and MAC tag. */
export async function signDump(key: CryptoKey, state: Uint8Array): Promise<Uint8Array> {
  // MAC and output are both built from the `withContext` copy, not `state`:
  // `state` may be a view over a transport buffer, and re-reading it after the
  // await could yield different bytes than were MAC'd
  const payload = withContext(state)
  const tag = await crypto.subtle.sign('HMAC', key, payload)
  const signed = new Uint8Array(1 + TAG_LEN + state.length)
  signed[0] = SIGNED_DUMP_VERSION
  signed.set(new Uint8Array(tag), 1)
  signed.set(payload.subarray(CONTEXT.length), 1 + TAG_LEN)
  return signed
}

/**
 * Verifies a signed dump and returns the inner worker envelope, throwing an
 * `Error` (whose message matches the Rust `PoolError::InvalidDump` display)
 * on any mismatch — callers must never send unverified bytes to a worker.
 */
export async function verifyDump(key: CryptoKey, signed: Uint8Array): Promise<Uint8Array> {
  if (signed.length < 1 + TAG_LEN) {
    throw new Error('invalid dump: too short to be a signed dump')
  }
  if (signed[0] !== SIGNED_DUMP_VERSION) {
    throw new Error(`invalid dump: unsupported signed-dump version ${signed[0]} (expected ${SIGNED_DUMP_VERSION})`)
  }
  const tag = copyBytes(signed.subarray(1, 1 + TAG_LEN))
  // the returned envelope is a view over the `withContext` copy, never the
  // caller's buffer: `signed` could be mutated during the verify await,
  // letting unverified bytes reach a worker
  const payload = withContext(signed.subarray(1 + TAG_LEN))
  const ok = await crypto.subtle.verify('HMAC', key, tag, payload)
  if (!ok) {
    throw new Error(
      'invalid dump: signature verification failed — the dump was signed with a different key or corrupted',
    )
  }
  return payload.subarray(CONTEXT.length)
}

/** `CONTEXT || state`, the exact bytes the MAC covers. */
function withContext(state: Uint8Array): Uint8Array<ArrayBuffer> {
  const buf = new Uint8Array(CONTEXT.length + state.length)
  buf.set(CONTEXT, 0)
  buf.set(state, CONTEXT.length)
  return buf
}

/** Returns `crypto.subtle`, throwing the standard dump-signing error when
 *  WebCrypto is unavailable (a browser page not in a secure context). */
function requireSubtle(): SubtleCrypto {
  if (typeof crypto === 'undefined' || crypto.subtle === undefined) {
    throw new Error('dump signing needs WebCrypto (crypto.subtle) — unavailable outside a secure context')
  }
  return crypto.subtle
}

/** Copies bytes into a fresh `ArrayBuffer`-backed array (WebCrypto rejects
 *  views over a possible `SharedArrayBuffer` at the type level). */
function copyBytes(bytes: Uint8Array): Uint8Array<ArrayBuffer> {
  const copy = new Uint8Array(bytes.length)
  copy.set(bytes)
  return copy
}

File: crates/monty-pool/src/dump_sign.rs
//! Host-side HMAC signing of dump payloads.
//!
//! [`crate::Checkout::dump`] bytes are persisted by callers and later fed back
//! to [`crate::Checkout::restore`], where they reach deserialization of
//! interpreter state inside a worker — so a forged or tampered dump is a real
//! attack surface. Every dump is therefore signed with the pool's [`DumpKey`]
//! and verified before a `Load` is ever sent to a worker. The key lives only
//! in the parent process; workers (which run untrusted code) never see it.
//!
//! Signed payload layout (`SIGNED_DUMP_VERSION` = 1):
//!
//! ```text
//! [version u8 = 0x01][tag: 32-byte HMAC-SHA256(key, CONTEXT || inner)][inner]
//! ```
//!
//! where `inner` is the worker's opaque dump envelope, untouched. `CONTEXT`
//! domain-separates the MAC so a dump key reused elsewhere cannot be tricked
//! into signing/verifying non-dump data. The TypeScript mirror for the wasm
//! worker path (`crates/monty-js/ts/worker/dumpSign.ts`) must stay
//! byte-compatible with this file — it is the source of truth for the format.

use std::{error, fmt};

use hmac::{Hmac, Mac};
use sha2::Sha256;

use crate::PoolError;

/// Minimum accepted [`DumpKey`] length; shorter keys are rejected at
/// construction so weak keys never make it into a pool.
pub const MIN_DUMP_KEY_LEN: usize = 16;

/// Format version prepended to every signed dump.
const SIGNED_DUMP_VERSION: u8 = 1;

/// Domain-separation prefix mixed into every MAC (see module docs).
const CONTEXT: &[u8] = b"monty-dump-sign-v1";

/// HMAC-SHA256 output length in bytes.
const TAG_LEN: usize = 32;

/// A pool's dump-signing key.
///
/// Supply one via [`crate::PoolConfig::dump_key`] to make dumps restorable
/// across pools/processes; without one the pool generates a random ephemeral
/// key on first use, so its dumps only restore into that same pool instance.
#[derive(Clone)]
pub struct DumpKey(Vec<u8>);

impl DumpKey {
    /// Wraps user-supplied key bytes, rejecting keys shorter than
    /// [`MIN_DUMP_KEY_LEN`].
    pub fn new(key: Vec<u8>) -> Result<Self, InvalidDumpKey> {
        if key.len() < MIN_DUMP_KEY_LEN {
            Err(InvalidDumpKey)
        } else {
            Ok(Self(key))
        }
    }

    /// Generates a random 32-byte key from OS randomness, used when
    /// [`crate::PoolConfig::dump_key`] is `None`. Created lazily on the first
    /// dump/restore, so pools that never dump draw no randomness. Panics if
    /// the OS RNG fails — an unrecoverable environment fault, not a pool error.
    pub(crate) fn ephemeral() -> Self {
        let mut key = vec![0u8; 32];
        getrandom::fill(&mut key).expect("OS randomness unavailable — cannot generate a dump signing key");
        Self(key)
    }

    /// Signs a worker dump envelope, prepending the version byte and MAC tag.
    pub(crate) fn sign(&self, state: &[u8]) -> Vec<u8> {
        let mut mac = self.mac();
        mac.update(state);
        let tag = mac.finalize().into_bytes();
        let mut signed = Vec::with_capacity(1 + TAG_LEN + state.len());
        signed.push(SIGNED_DUMP_VERSION);
        signed.extend_from_slice(&tag);
        signed.extend_from_slice(state);
        signed
    }

    /// Verifies a signed dump and returns the inner worker envelope, or
    /// [`PoolError::InvalidDump`] — the caller must not send unverified bytes
    /// to a worker. The tag comparison is constant-time (`Mac::verify_slice`).
    pub(crate) fn verify(&self, mut signed: Vec<u8>) -> Result<Vec<u8>, PoolError> {
        if signed.len() < 1 + TAG_LEN {
            return Err(PoolError::InvalidDump("too short to be a signed dump".to_owned()));
        }
        if signed[0] != SIGNED_DUMP_VERSION {
            return Err(PoolError::InvalidDump(format!(
                "unsupported signed-dump version {} (expected {SIGNED_DUMP_VERSION})",
                signed[0]
            )));
        }
        let inner = signed.split_off(1 + TAG_LEN);
        let mut mac = self.mac();
        mac.update(&inner);
        match mac.verify_slice(&signed[1..]) {
            Ok(()) => Ok(inner),
            Err(_) => Err(PoolError::InvalidDump(
                "signature verification failed — the dump was signed with a different key or corrupted".to_owned(),
            )),
        }
    }

    /// A MAC instance keyed with this key, pre-fed with the domain-separation
    /// context.
    fn mac(&self) -> Hmac<Sha256> {
        let mut mac = Hmac::<Sha256>::new_from_slice(&self.0).expect("HMAC-SHA256 accepts keys of any length");
        mac.update(CONTEXT);
        mac
    }
}

/// Redacted — key material must never reach logs (`PoolConfig` derives Debug).
impl fmt::Debug for DumpKey {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("DumpKey(..)")
    }
}

/// Error from [`DumpKey::new`]: the supplied key is shorter than
/// [`MIN_DUMP_KEY_LEN`].
#[derive(Debug)]
pub struct InvalidDumpKey;

impl fmt::Display for InvalidDumpKey {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "dump key must be at least {MIN_DUMP_KEY_LEN} bytes")
    }
}

impl error::Error for InvalidDumpKey {}
